enhancements/network/upi-node-ip-selection.md

```diff
@@ -0,0 +1,216 @@
+- Users can specify the network interface at installation at
+  install-config.yaml.
+
+### Non-Goals
+
+- Support IPI platforms
+
+## Proposal
+
+Enable the nodeip-configuration service for all the UPI platforms. It will be
+the single source for the node ip/interface selection.
+
+### Workflow Description
+
+- At deployment time the cluster creator can specify the `machineNetwork` in the
+  install-config.yaml.
+
+- Openshift installer set the `networking.machineNetwork` field of the
+  `infrastructures.config.openshift.io/cluster` resource.
+
+- Machine config operator renders the nodeip-configuration.service and set
+  `NODEIP_HINT` to the value of `machineNetwork`.
+
+### API Extensions
+
+Enhancement requires below modifications to the mentioned CRDs
+- Add `networking.machineNetwork` field to `spec` of the
+  `infrastructures.config.openshift.io`
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: infrastructures.config.openshift.io
+spec:
+  versions:
+  - name: v1
+    schema:
+      openAPIV3Schema:
+        properties:
+          spec:
+            properties:
+              networking:
+                properties:
+                  machineNetwork:
+                    description: MachineNetwork is the list of IP address pools for machines.
+                    items:
+                      description: MachineNetworkEntry is a single IP address block for
+                        node IP blocks.
+                      properties:
+                        cidr:
+                          description: CIDR is the IP block address pool for machines
+                            within the cluster.
+                          type: Any
+                      required:
+                      - cidr
+                      type: object
+                    type: array
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+
+### Risks and Mitigations
+
+- Currently all host services are expected to listen on the same IP and
+  interface. If at some point in the future we need host services listening
+  on multiple different interfaces, this may not work. However, because we
+  are centralizing all IP selection logic in nodeip-configuration, it should
+  be possible to extend that to handle multiple interfaces if necessary.
+
+### Drawbacks
+
+N/A
+
+## Design Details
+
+- Ask users to specify the `machineNetwork` in the install-config.yaml for UPI
+  platforms when there are multiple NICs on the hosts.
+
+- The installer shall populate the value of the `machineNetwork` to the spec
+  of the `infrastructures.config.openshift.io/cluster` resource
+
+- Enable the nodeip-configuration.service for all the UPI platforms in the
+  machine config operator. When MCO renders the nodeip-configuration.service, it
+  shall set the `NODEIP_HINT` variable with the value of the `machineNetwork`.
+
+- Update the nodeip-configuration.service script to filter out the node subnet
```

Comment:
I don't understand this point.  Passing machineNetwork as the hint will force runtimecfg to select the IP from that subnet. Why would we filter anything further?

Also, I'd much prefer to have any filtering happen in runtimecfg since doing things in templates or scripts makes them hard to test.

Reply:
The `machineNetwork` is a CIDR list. Is that possible to have multiple CIDRs for one address family? Can we assume it only contains one v4 or v6 CIRD at most? If the answer is yes, I think we can remove this line.

Reply:
Oh, right. That could be a problem. In some platforms we aren't going to know which CIDR needs to be used ahead of time. Maybe NODEIP_HINT needs to take a list so runtimecfg can consider all of them in order when selecting an IP? That would be a significant change in how it works, but I'm not sure how else we handle this. At least it could be backward compatible.